### PR TITLE
Cleanup unused properties in rules

### LINKED ISSUE #2966: PERF: Clean up unused properties in file rules
File rules are used to produce the project tree through dataflow.  Currently, the project system often reuse browse object rules with file rules, which adds extra properties, which are evaluated, but never used by any feature.  This adds overhead to CPU/memory usages.

We want to get rid of those properties in file context rules (Compile/None/Content/AdditionalFiles/EmbeddedResource)

Identity FileNameAndExtension URL Extension.

The impact is bigger for rules covering lots of files in the project, so focus on those rules first.

### PR DESCRIPTION
**Pull request by Pilchie**
Currently CPS does a bunch of work to make these available, even though no one uses them.

Fixes #2966.

<details>**Customer scenario**

What does the customer do to get into this situation, and why do we think this
is common enough to address in Escrow.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

**Bugs this fixes:** 

(either VSO or GitHub links)

**Workarounds, if any**

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

**Risk**

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

**Performance impact**

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

**Is this a regression from a previous update?**

**Root cause analysis:**

How did we miss it?  What tests are we adding to guard against it in the future?

**How was the bug found?**

(E.g. customer reported it vs. ad hoc testing)
</details>

### CODE REVIEW

**Comment by Pilchie** [comment]
I think this breaks stuff in the dependencies node, but I'll investigate later.  @lifengl

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ResolvedCOMReference.xaml [hunk 1]
```diff
@@ -66,15 +66,6 @@
         <EnumValue Name="Native Assembly" DisplayName="Native Assembly" />
     </EnumProperty>
 
-    <StringProperty Name="Identity"
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ResolvedCOMReference.xaml]: I don't think @lifengl was referring to any of the rules that represent "target results". These aren't used in the "file context" as called out in the bug.

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ResolvedCOMReference.xaml]: You can see the context here: https://github.com/dotnet/project-system/blob/ee188a4b5c4d03c63f10f461810e7352e6cf85db/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.Managed.DesignTime.targets#L164.

Comment (Pilchie) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ResolvedCOMReference.xaml]: Yeah, I wasn't sure about these - just figured that I might as well do all of the ones that were easy to do.  These could be the source dependencies node issues though.

**Comment by lifengl** [comment]
This change might break some features, because those rules are used to create browse and as file rules to produce the tree.  Some of those properties might be needed in a browser object and may be the reason there.  It is ok to have more properties in a browse rule, but a file rule is repeatedly evaluated and results are kept in memory, so to reduce memory used by file rules, it would require to separate two rules first.

A quick partial fix might be to fix the rule for csharp file first.  We have already had two different rules for CS file.

Sent from my phone

On Jan 3, 2018, at 7:31 PM, Kevin Pilch <notifications@github.com<mailto:notifications@github.com>> wrote:


@Pilchie commented on this pull request.

________________________________

In src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ResolvedCOMReference.xaml<https://nam03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fproject-system%2Fpull%2F3100%23discussion_r159575924&data=02%7C01%7C%7C19fe0d0022e84d8ac2a808d55323b399%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636506335156152687&sdata=qZHVbHN5hmDmX%2FXwhCEcx1nd4FRVLNvURmXjdJLbhs4%3D&reserved=0>:

> @@ -66,15 +66,6 @@
         <EnumValue Name="Native Assembly" DisplayName="Native Assembly" />
     </EnumProperty>

-    <StringProperty Name="Identity"


Yeah, I wasn't sure about these - just figured that I might as well do all of the ones that were easy to do. These could be the source dependencies node issues though.

**Comment by Pilchie** [comment]
`Compile` already doesn't have any of these properties (it's not changed by this PR), but maybe `None` and `Content` should be done?

**Comment by davkean** [comment]
We should avoid changing any that have a context of "BrowseObject" (Properties window) or any rule that are "target results".

**Comment by davkean** [comment]
This is actually harder than just removing the properties; we have file and browse object  ("Properties" windows and Automation APIs) that share the same rule file. Some of those properties are available via the automation APIs and bring compat with legacy project system, I think we'll need to split the rules. Just trying to figure out what APIs the legacy project system provides here to confirm.

**Comment by davkean** [comment]
Closing in lieu of https://github.com/dotnet/project-system/pull/3484.

**Pull request closed without merging**